### PR TITLE
SD-354: Support Delegated Authentication for Sync

### PR DESCRIPTION
**Pull request by Fred-Simeon**
This PR adds support to the Simeon Installer for delegated authentication which was requested in SD-354.

### CODE REVIEW

File: SimeonInstaller.ps1 [hunk 1]
```diff
@@ -1605,12 +1631,14 @@ CRLFOption=CRLFAlways
...
             }
 
+            $pipelineId = $pipeline.id
             if ($pipeline) {
                 Write-Information "Pipeline '$pipelineName' already exists - updating"
 
                 $definition = irm @restProps "$apiBaseUrl/build/definitions/$($pipeline.id)?revision=$($pipeline.revision)" -Method Get
 
                 $body.variables = $definition.variables
+                if (!$Credential) { $body.variables | gm |? Name -in @('AadAuth:Username', 'AadAuth:Password') | % {    $body.PSObject.Properties.Remove($_.Name)   } }
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1641]: add 'AadAuth:TenantId' and 'AzureManagement:SubscriptionId' here...an alternative I thought of:
```
        if (!$UseServiceAccount) {
            $pipelineVariables['AadAuth:TenantId'] = @{
                allowOverride = $false
                value = $Tenant
            }
        }
```
set them to null here...your call

**Review by jeff-simeon — COMMENTED** [review]
nice job

File: SimeonInstaller.ps1 [hunk 2]
```diff
@@ -662,6 +663,32 @@ CRLFOption=CRLFAlways
...
     }
 
+    <#
+    .SYNOPSIS
+    Removes a service account named simeon@yourcompany.com
+    #>
+    function Remove-SimeonTenantServiceAccount {
+        [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSReviewUnusedParameter', '', Scope = 'Function')]
+        [CmdletBinding()]
+        [OutputType([pscredential])]
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:673]: There is no output to this function is there?

File: SimeonInstaller.ps1 [hunk 3]
```diff
@@ -1605,12 +1632,14 @@ CRLFOption=CRLFAlways
...
             }
 
+            $pipelineId = $pipeline.id
             if ($pipeline) {
                 Write-Information "Pipeline '$pipelineName' already exists - updating"
 
                 $definition = irm @restProps "$apiBaseUrl/build/definitions/$($pipeline.id)?revision=$($pipeline.revision)" -Method Get
 
                 $body.variables = $definition.variables
+                if (!$Credential) { $body.variables | gm |? Name -in @('AadAuth:Username', 'AadAuth:Password', 'AadAuth:TenantId', 'AzureManagement:SubscriptionId') | % {    $body.variables.PSObject.Properties.Remove($_.Name)   } }
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1642]: Looks like some extra spaces. Please auto format

File: SimeonInstaller.ps1 [hunk 4]
```diff
@@ -1605,12 +1632,14 @@ CRLFOption=CRLFAlways
                 variables = $pipelineVariables
             }
 
+            $pipelineId = $pipeline.id
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1635]: Why is this added?

File: SimeonInstaller.ps1 [hunk 5]
```diff
@@ -1637,7 +1666,82 @@ CRLFOption=CRLFAlways
...
+                $secureFileId = $secureFile.id
+                Write-Information "Created secure file $secureFileId"
+                Write-Information "Setting secure file $secureFileId to be accessible by pipeline $($pipelineId)"
+                Invoke-WithRetry { Invoke-RestMethod @restProps "$apiBaseUrl/pipelines/pipelinePermissions/securefile/$secureFileId" -Method Patch -Body (@{
+                    resources = @{}
+                    pipelines = @(@{ authorized = $true; id = $($pipelineId) })
+                } | ConvertTo-Json -Depth 100) | Out-Null }
+
+                # Set Role Assignment
+                foreach ($user in @("Tenants Build Service", "Project Collection Build Service")) {
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1711]: It will the “$Project Build Service”. Don’t assume “Tenants”

File: SimeonInstaller.ps1 [hunk 6]
```diff
@@ -1637,7 +1666,82 @@ CRLFOption=CRLFAlways
...
+                $pipelineId = $buildDefinition.id
+            }
+
+            if (!$Credential) {
+                $cachePath = Join-Path ([IO.Path]::GetTempPath()) 'M365Management'
+                if (!(Test-Path $cachePath)) {
+                    New-Item -Path $cachePath -ItemType "directory" -Force | Out-Null
+                }
+                # remove unnecessary cache files
+                gci $cachePath -Exclude @('keys', 'msal') | Remove-Item -Recurse -Force
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1679]: This will never exist in this context, will it?

File: SimeonInstaller.ps1 [hunk 7]
```diff
@@ -1637,7 +1666,82 @@ CRLFOption=CRLFAlways
...
+                            "options": {
+                                "MinResults": 1,
+                                "MaxResults": 20
+                            },
+                            "properties": [
+                                "DisplayName"
+                            ]
+                        }
+"@ | Out-Null }
+                    $contributorsDisplayName = "$user ($Organization)"
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1733]: Why is this variable called contributors?

**Review by jeff-simeon — COMMENTED** [review]
This is good. The one thing that bothers me a tiny bit is that we lose the ability to update a pipeline using a service account without having to recreate the service account (consider a situation where we need to update everyone’s pipelines but don’t want to change the variables). If you can think of an easy answer, let me know

File: SimeonInstaller.ps1 (hunk 4 above)
Comment (Fred-Simeon) [reply SimeonInstaller.ps1:1635]: What do you mean?

**Review by jeff-simeon — APPROVED** [review]

File: SimeonInstaller.ps1 [hunk 8]
```diff
@@ -1723,9 +1722,10 @@ CRLFOption=CRLFAlways
                                 "DisplayName"
                             ]
                         }
-"@ | Out-Null }
+"@ }
                     $userDisplayName = "$user ($Organization)"
                     $userId = $identities.results.identities |? displayName -eq $userDisplayName | Select -ExpandProperty localId
+                    Write-Information "Making $userDisplayName ($userId) admin for Secure File"
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1728]: Print the id please

File: SimeonInstaller.ps1 [hunk 9]
```diff
@@ -1637,7 +1666,76 @@ CRLFOption=CRLFAlways
             }
             else {
                 Write-Information "Creating pipeline '$pipelineName'"
-                irm @restProps "$apiBaseUrl/build/definitions" -Method Post -Body ($body | ConvertTo-Json -Depth 10) | Out-Null
+                $buildDefinition = Invoke-WithRetry { irm @restProps "$apiBaseUrl/build/definitions" -Method Post -Body ($body | ConvertTo-Json -Depth 10) }
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1669]: Why use a different variable name than $pipeline?

File: SimeonInstaller.ps1 [hunk 10]
```diff
@@ -1637,7 +1666,76 @@ CRLFOption=CRLFAlways
...
+
+                Write-Information "Creating secure file $secureFileName"
+                $createSecureFileUri = "$secureFilesUri`?name=$([System.Net.WebUtility]::UrlEncode($secureFileName))"
+                $secureFileCreateRestProps = $restProps.Clone()
+                $secureFileCreateRestProps['ContentType'] = 'application/octet-stream'
+                $secureFile = Invoke-WithRetry { (Invoke-RestMethod @secureFileCreateRestProps $createSecureFileUri -Method Post -InFile "$cachePath.zip") }
+                $secureFileId = $secureFile.id
+                Write-Information "Created secure file $secureFileId"
+
+                $pipelineId = $pipeline ? $pipeline.id : $buildDefinition.id
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1696]: Why use a different variable name than $pipeline?

File: SimeonInstaller.ps1 (hunk 8 above)
Comment (jeff-simeon) [reply SimeonInstaller.ps1:1728]: Lower case “secure file”

File: SimeonInstaller.ps1 [hunk 11]
```diff
@@ -1803,6 +1902,17 @@ CRLFOption=CRLFAlways
                 if (Test-Path "$_.yml") { Remove-Item "$_.yml" -Force -EA SilentlyContinue }
                 irm "https://raw.githubusercontent.com/simeoncloud/$ymlRepo/master/$_.yml" -OutFile "$_.yml"
             }
+
+            if (!$UseServiceAccount) {
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1906]: Should we remove the use cache parameter if using a service account?

File: SimeonInstaller.ps1 [hunk 12]
```diff
@@ -1868,11 +1986,18 @@ CRLFOption=CRLFAlways
         $pipelineVariables = @{}
         if ($Subscription) {
             $pipelineVariables['AzureManagement:SubscriptionId'] = @{
-                allowOverride = $true
+                allowOverride = $false
```
Comment (jeff-simeon) [inline SimeonInstaller.ps1:1989]: Are there other usages of allowOverride? If so, let’s default them to false too

**Review by jeff-simeon — COMMENTED** [review]
Please make sure any script analyzer problems are resolved too

File: SimeonInstaller.ps1 (hunk 8 above)
Comment (Fred-Simeon) [reply SimeonInstaller.ps1:1728]: > Print the id please
The id of the secure file?

Comment (jeff-simeon) [reply SimeonInstaller.ps1:1728]: Yes just like in the yml

File: SimeonInstaller.ps1 (hunk 11 above)
Comment (Fred-Simeon) [reply SimeonInstaller.ps1:1906]: I suppose we could. Do you think we should do that?

Comment (jeff-simeon) [reply SimeonInstaller.ps1:1906]: Yes

Comment (jeff-simeon) [reply SimeonInstaller.ps1:1906]: Otherwise If you reinstall with use service account checked, it will still use the cache and not authenticate as the service account

**Review by jeff-simeon — APPROVED** [review]

**Review by david-simeon — APPROVED** [review]

**Review by jeff-simeon — APPROVED** [review]
